src/librustc_driver/driver.rs

```diff
@@ -64,6 +64,69 @@ use pretty::ReplaceBodyWithLoop;
 
 use profile;
 
+#[cfg(not(parallel_queries))]
+pub fn spawn_thread_pool<F: FnOnce(config::Options) -> R + sync::Send, R: sync::Send>(
+    opts: config::Options,
+    f: F
+) -> R {
+    f(opts)
+}
+
+#[cfg(parallel_queries)]
+pub fn spawn_thread_pool<F: FnOnce(config::Options) -> R + sync::Send, R: sync::Send>(
+    opts: config::Options,
+    f: F
+) -> R {
+    use syntax;
+    use syntax_pos;
+    use scoped_tls::ScopedKey;
+    use rayon::{ThreadPoolBuilder, ThreadPool};
+
+    let config = ThreadPoolBuilder::new().num_threads(Session::query_threads_from_opts(&opts))
+                                         .stack_size(16 * 1024 * 1024);
+
+    let with_pool = move |pool: &ThreadPool| {
+        pool.install(move || f(opts))
+    };
+
+    fn try_with<T, F, R>(key: &'static ScopedKey<T>, f: F) -> R
+        where F: FnOnce(Option<&T>) -> R
+    {
+        if key.is_set() {
+            key.with(|v| f(Some(v)))
+        } else {
+            f(None)
+        }
+    }
+
+    fn maybe_set<T, F, R>(key: &'static ScopedKey<T>,
+                          value: Option<&T>, f: F) -> R
+        where F: FnOnce() -> R
+    {
+        if let Some(v) = value {
+            key.set(v, f)
+        } else {
+            f()
+        }
+    }
+
+    try_with(&syntax::GLOBALS, |syntax_globals| {
```

Comment:
It took me a while to see what this was doing. Now I grok that it is basically just copying the values over from this thread into the subthreads -- one question: under what conditions would these values be `None`? Can that ever happen?

Reply:
(Also, a comment explaining what is going on would be good)

Reply:
I remember there were conditions where those values were `None`, I can't remember how that happened though. It does not make sense to run the compiler without a `Symbol` interner.

Reply:
I've added a comment and removed the code dealing with the `None` case.